File: Source/Binary/Creators.External/UnionConverter.cs
using Mikodev.Binary.Internal.Delegates;
using System;
using System.Diagnostics;
using System.Runtime.CompilerServices;

namespace Mikodev.Binary.Creators.External
{
    internal sealed class UnionConverter<T> : Converter<T>
    {
        private const int MarkNone = 0;

        private readonly bool noNull;

        private readonly OfUnion<T> ofUnion;

        private readonly ToUnion<T> toUnion;

        private readonly OfUnion<T> ofUnionWith;

        private readonly ToUnion<T> toUnionWith;

        public UnionConverter(OfUnion<T> ofUnion, ToUnion<T> toUnion, OfUnion<T> ofUnionWith, ToUnion<T> toUnionWith, bool noNull) : base(0)
        {
            this.noNull = noNull;
            this.ofUnion = ofUnion;
            this.toUnion = toUnion;
            this.ofUnionWith = ofUnionWith;
            this.toUnionWith = toUnionWith;
        }

        [DebuggerStepThrough, MethodImpl(MethodImplOptions.NoInlining)]
        private void ThrowNull() => throw new ArgumentNullException("item", $"Union can not be null, type: {ItemType}");

        [DebuggerStepThrough, MethodImpl(MethodImplOptions.NoInlining)]
        private void ThrowInvalid(int mark) => throw new ArgumentException($"Invalid union tag '{mark}', type: {ItemType}");

        [DebuggerStepThrough, MethodImpl(MethodImplOptions.AggressiveInlining)]
        private void ThrowOnNull(T item) { if (noNull && item == null) ThrowNull(); }

        [DebuggerStepThrough, MethodImpl(MethodImplOptions.AggressiveInlining)]
        private void ThrowOnInvalid(int mark) { if (mark != MarkNone) ThrowInvalid(mark); }

        public override void ToBytes(ref Allocator allocator, T item)
        {
            ThrowOnNull(item);
            var mark = MarkNone;
            ofUnion.Invoke(ref allocator, item, ref mark);
            ThrowOnInvalid(mark);
        }

        public override T ToValue(in ReadOnlySpan<byte> span)
        {
            var temp = span;
            var mark = MarkNone;
            var item = toUnion.Invoke(ref temp, ref mark);
            ThrowOnInvalid(mark);
            return item;
        }

        public override void ToBytesWithMark(ref Allocator allocator, T item)
        {
            ThrowOnNull(item);
            var mark = MarkNone;
            ofUnionWith.Invoke(ref allocator, item, ref mark);
            ThrowOnInvalid(mark);
        }

        public override T ToValueWithMark(ref ReadOnlySpan<byte> span)
        {
            var mark = MarkNone;
            var item = toUnionWith.Invoke(ref span, ref mark);
            ThrowOnInvalid(mark);
            return item;
        }
    }
}

File: code/Binary.FSharp/Creators/UnionConverter.fs
namespace Mikodev.Binary.Creators

open Mikodev.Binary
open System
open System.Diagnostics
open System.Runtime.CompilerServices

type internal UnionEncoder<'T> = delegate of allocator: byref<Allocator> * item: 'T * mark: byref<int> -> unit

type internal UnionDecoder<'T> = delegate of span: byref<ReadOnlySpan<byte>> * mark: byref<int> -> 'T

type internal UnionConverter<'T>() =
    inherit Converter<'T>(0)

    [<Literal>]
    let constant = 0

    [<DefaultValue>]
    val mutable encode: UnionEncoder<'T>

    [<DefaultValue>]
    val mutable encodeAuto: UnionEncoder<'T>

    [<DefaultValue>]
    val mutable decode: UnionDecoder<'T>

    [<DefaultValue>]
    val mutable decodeAuto: UnionDecoder<'T>

    [<DefaultValue>]
    val mutable needNullCheck: bool

    [<DebuggerStepThrough>]
    member private __.ExceptNull() : unit =
        raise (ArgumentNullException("item", $"Union can not be null, type: {typeof<'T>}"))

    [<DebuggerStepThrough>]
    member private __.ExceptMark(mark: int) : unit = raise (ArgumentException $"Invalid union tag '{mark}', type: {typeof<'T>}")

    [<DebuggerStepThrough>]
    [<MethodImpl(MethodImplOptions.AggressiveInlining)>]
    member private me.HandleNull(item: 'T) : unit =
        if me.needNullCheck && isNull (box item) then
            me.ExceptNull()
        ()

    [<DebuggerStepThrough>]
    [<MethodImpl(MethodImplOptions.AggressiveInlining)>]
    member private me.HandleMark(mark: int) : unit =
        if mark <> constant then
            me.ExceptMark mark
        ()

    member me.Initialize(encode: UnionEncoder<'T>, encodeAuto: UnionEncoder<'T>, decode: UnionDecoder<'T>, decodeAuto: UnionDecoder<'T>, needNullCheck: bool) =
        assert (isNull me.encode)
        assert (isNull me.encodeAuto)
        me.encode <- encode
        me.encodeAuto <- encodeAuto
        me.decode <- decode
        me.decodeAuto <- decodeAuto
        me.needNullCheck <- needNullCheck
        ()

    override me.Encode(allocator, item) =
        me.HandleNull item
        let mutable mark = constant
        me.encode.Invoke(&allocator, item, &mark)
        me.HandleMark mark
        ()

    override me.EncodeAuto(allocator, item) =
        me.HandleNull item
        let mutable mark = constant
        me.encodeAuto.Invoke(&allocator, item, &mark)
        me.HandleMark mark
        ()

    override me.Decode(span: inref<ReadOnlySpan<byte>>) : 'T =
        let mutable body = span
        let mutable mark = constant
        let item = me.decode.Invoke(&body, &mark)
        me.HandleMark mark
        item

    override me.DecodeAuto span =
        let mutable mark = constant
        let item = me.decodeAuto.Invoke(&span, &mark)
        me.HandleMark mark
        item
